File: Back/NodeJS/modules/bets/src/infrastructure/repositories/sql/SqlBetRepository.ts
import { DomainEventAccessor } from "../../../../../shared/infrastructure/events/DomainEventAccessor";
import { Bet } from "../../../domain/bets/Bet";
import { BetId } from "../../../domain/bets/BetId";
import { IBetRepository } from "../../../domain/bets/IBetRepository";
import { MemberId } from "../../../domain/members/MemberId";
import { BetContext } from "../../DataAccess/BetContext";
import { BetEntity } from "../../DataAccess/entities/BetEntity";

export class SqlBetRepository implements IBetRepository {

    constructor(private readonly betContext: BetContext, private readonly domainEventsAccessor: DomainEventAccessor){}

    async GetById(betId: BetId): Promise<Bet | undefined> {
        const betEntity = await this.betContext.BetEntity.findOneBy({
            BetId: betId.Value
        });
        if(!betEntity) {
            return undefined!
        }
        return Bet.CreateFromEntity(new BetId(betEntity.BetId),
                                    new MemberId(betEntity.BettorId),
                                    betEntity.Coins,
                                    betEntity.Description,
                                    betEntity.EndDate,
                                    betEntity.MaxAnswerDate,
                                    betEntity.Guests.split(";"),
                                    betEntity.IsSuccessful);
    }
    async Save(bet: Bet): Promise<void> {
        const entity = BetEntity.Create(bet);
        await this.betContext.BetEntity.save(entity);
        this.domainEventsAccessor.AddEvents(bet.DomainEvents)
    }

}

File: Back/.Net/Modules/Bets/BetFriends.Bets.Infrastructure/Repositories/Sql/SqlBetRepository.cs
using BetFriends.Bets.Domain.Bets;
using BetFriends.Bets.Infrastructure.Repositories.Sql.DataAccess;
using BetFriends.Shared.Domain;
using BetFriends.Shared.Infrastructure.Event;

namespace BetFriends.Bets.Infrastructure.Repositories.Sql;

public class SqlBetRepository(BetContext betContext, DomainEventsAccessor domainEventsAccessor) : IBetRepository
{
    private readonly BetContext betContext = betContext;
    private readonly DomainEventsAccessor domainEventsAccessor = domainEventsAccessor;

    public async Task<Bet> GetByIdAsync(BetId betId)
    {
        var betEntity = await betContext.Bets.FindAsync(betId.Value);
        if (betEntity == null)
            return default!;
        return Bet.CreateFromEntity(betEntity.Id,
                                    betEntity.BettorId,
                                    betEntity.Description,
                                    betEntity.Coins,
                                    betEntity.EndDate,
                                    betEntity.Guests,
                                    betEntity.MaxAnswerDate,
                                    betEntity.IsSuccessful);
    }

    public BetEntity? GetEntity(Guid betId)
    {
        return betContext.Bets.Find(betId);
    }

    public async Task SaveAsync(Bet bet)
    {
        var entity = new BetEntity(bet);
        await betContext.AddAsync(entity);
        domainEventsAccessor.Add(bet.Events);
    }

    public void SaveEntity(BetEntity betEntity)
    {
        betContext.Add(betEntity);
    }
}
